kallsyms/kallsyms.go

```diff
@@ -0,0 +1,667 @@
+		default:
+			if mod.start == 0 {
+				mod.start = libpf.Address(address)
+			}
+			if addr := libpf.Address(address); addr >= mod.start && addr < mod.end {
+				// Add symbol to the module symbols
+				mod.symbols = append(mod.symbols, symbol{
+					offset: uint32(addr - mod.start),
+					index:  mod.addName(fields[2]),
+				})
+			}
+		}
+	}
+	if mod != nil {
+		mod.finish()
+	}
+	if noSymbols {
+		return ErrSymbolPermissions
+	}
+
+	sort.Slice(mods, func(i, j int) bool {
+		return mods[i].start >= mods[j].start
+	})
+	// Heap allocate the exact amount needed. This also makes the initial
+	// buffer stack allocated.
+	s.modules.Store(slices.Clone(mods))
+	return nil
+}
+
+// loadKallsyms will reload kernel symbols. This function can run concurrently with
+// module module lookups. The reload result is visible atomically after success.
+func (s *Symbolizer) loadKallsyms() error {
+	file, err := os.Open("/proc/kallsyms")
+	if err != nil {
+		return fmt.Errorf("unable to open kallsyms: %v", err)
+	}
+	defer file.Close()
+
+	return s.updateSymbolsFrom(file)
+}
+
+var nonsyfsModules = libpf.Set[string]{
+	Kernel: libpf.Void{},
+	"bpf":  libpf.Void{},
+}
+
+// loadModules will reload module metadata.
+func (s *Symbolizer) loadModules() (bool, error) {
+	dir, err := os.Open(sysModule)
+	if err != nil {
+		return false, err
+	}
+	defer dir.Close()
+
+	needReloadSymbols := false
+	modules, _ := s.modules.Load().([]Module)
+	mods := make([]Module, 0, 400)
+
+	// Copy the modules not present in sysfs
+	for _, mod := range modules {
+		if _, ok := nonsyfsModules[mod.Name()]; ok {
+			mods = append(mods, mod)
+		}
+	}
+
+	// Scan sysfs for current module listing and its metadata
+	for {
+		dirEntries, err := dir.ReadDir(64)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return false, err
+		}
+		for _, dirEnt := range dirEntries {
+			if !dirEnt.IsDir() {
+				continue
+			}
+
+			moduleName := dirEnt.Name()
+			curMod := Module{}
+			if !loadModuleMetadata(&curMod, moduleName, 0) {
+				// sysfs contains directories also for statically built
+				// kernel modules. Ignore these.
+				continue
+			}
+
+			oldMod, _ := getModuleByAddress(modules, curMod.start)
+			if oldMod != nil && oldMod.Name() == moduleName && oldMod.mtime == curMod.mtime {
+				// Reuse the old module
+				mods = append(mods, *oldMod)
+			} else {
+				// Create a stub module without symbols
+				curMod.setStub(moduleName)
+				mods = append(mods, curMod)
+				needReloadSymbols = true
+			}
+		}
+	}
+
+	sort.Slice(mods, func(i, j int) bool {
+		return mods[i].start >= mods[j].start
+	})
+	// Heap allocate the exact amount needed. This also makes the initial
+	// buffer stack allocated.
+	s.modules.Store(slices.Clone(mods))
+
+	return needReloadSymbols, nil
+}
+
+// reloadWorker is the goroutine handling the reloads of the kallsyms.
+func (s *Symbolizer) reloadWorker(ctx context.Context, kobjectClient *kobject.Client) {
+	noTimeout := make(<-chan time.Time)
```

Comment:
Prevent channel leakage:
```suggestion
	noTimeout := make(<-chan time.Time)
	defer close(noTimeout)
```